keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+}
+
+type ClusterNetworkPolicyRule struct {
+	AppliedTo    AppliedTo
+	Ingress      []ClusterNetworkPolicyIngressRule
+	Egress       []ClusterNetworkPolicyEgressRule
+}
+
+type ClusterNetworkPolicyIngress/EgressRule struct {
+	// Action specifies whether this rule must pass, allow traffic or deny traffic.
+	// Required field for any rule.
+	Action       RuleAction
+	// List of ports for incoming/outgoing traffic.
+	// Each item in this list is combined using a logical OR. If this field is
+	// empty or missing, this rule matches all ports (traffic not restricted by port).
+	// If this field is present and contains at least one item, then this rule allows/denies
+	// traffic only if the traffic matches at least one port in the list.
+	// +optional
+	Ports        []networkingv1.NetworkPolicyPort
+	// List of sources/dest which should be able to access the pods selected for this rule.
+	// Items in this list are combined using a logical OR operation. If this field is
+	// empty or missing, this rule matches all sources/dest (traffic not restricted by
+	// source/dest). If this field is present and contains at least one item, this rule
+	// allows/denies traffic only if the traffic matches at least one item in the from/to list.
+	// +optional
+	From/To      []ClusterNetworkPolicyPeer
+	// Name is an identifier for this rule. Must be unique within an instance of CNP. Can be
+	// same across policy instances.
+	Name         string
+}
+
+// WorkloadSelector selects workloads based on the label Selector.
+// Right now the only selector type supported is PodSelector. More selector types may be
+// added in the future to support other workload selection mechanism, e.g. serviceAccountSelector.
+// Must be accompanied by a NamespaceSelector to specify the scope of the workload selection.
+type WorkloadSelector struct {
+	PodSelector         *metav1.LabelSelector
+}
+```
+
+**Alternatives considered for the WorkloadSelector struct:**  
+```golang
+type WorkloadSelector struct {
+	selectorType        WorkloadSelectorType
+	selector            metav1.LabelSelector
+}
+
+type WorkloadSelectorType string
+
+// Add future selector types here to ease introduction of future selectors (e.g. serviceAccountSelector)
+// and ensure api fails closed cleanly if a CNI doesn't support specific types of selectors.
+const (
+	PodSelector WorkloadSelectorType  = "PodSelector"
+)
+```
+This alternative WorkloadSelector definition is more explicit but more verbose as well.
+At this point, it is expected that any future selectors will be mutually exclusive to the
+PodSelector in a single WorkloadSelector. Both approaches should be future extensible
+without breaking the premise of fail-closed API design.
+
+(TODO) Does there seem to be a valid usecase where two or more types of selectors will
+need to co-exist in a same workloadSelector? i.e. select workloads in these Namespaces,
+where it matches certain Pod labels, AND matches serviceAccount selection criteria?
```

Comment:
Your user stories should provide the answer to this question. Right now they don't provide any use case for either podSelector or serviceAccountSelector.